iohk/NixOps.hs

```diff
@@ -0,0 +1,531 @@
+nixops' o c (NixopsCmd com) args = cmd' o "nixops" (com : nixopsCmdOptions o c <> args)
+
+
+-- * Deployment lifecycle
+--
+exists :: Options -> NixopsConfig -> IO Bool
+exists o c@NixopsConfig{..} = do
+  (code, _) <- nixops' o c "info" []
+  pure $ code == ExitSuccess
+
+create :: Options -> NixopsConfig -> IO ()
+create o c@NixopsConfig{..} = do
+  clusterExists <- exists o c
+  when clusterExists $
+    die $ format ("Cluster already exists?: '"%s%"'") cName
+  printf ("Creating cluster "%s%"\n") cName
+  nixops o c "create" $ deploymentFiles cEnvironment cTarget cElements
+
+modify :: Options -> NixopsConfig -> IO ()
+modify o c@NixopsConfig{..} = do
+  printf ("Syncing Nix->state for cluster "%s%"\n") cName
+  nixops o c "modify" $ deploymentFiles cEnvironment cTarget cElements
+
+deploy :: Options -> NixopsConfig -> IO ()
+deploy o c@NixopsConfig{..} = do
+  when (elem Nodes cElements) $ do
+     keyExists <- testfile "keys/key1.sk"
+     unless keyExists $
+       die "Deploying nodes, but 'keys/key1.sk' is absent."
+
+  printf ("Deploying cluster "%s%"\n") cName
+  export "GC_INITIAL_HEAP_SIZE" (showT $ 8 * 1024*1024*1024) -- for 100 nodes it eats 12GB of ram *and* needs a bigger heap
+  export "SMART_GEN_IP"     =<< incmd o "curl" ["--silent", fromURL awsPublicIPURL]
+
+  when (elem Explorer cElements) $ do
+    cmd o "scripts/generate-explorer-frontend.sh" []
+
+  nixops o c "set-args" $ concat $ uncurry nixArgCmdline <$> Map.toList cDeplArgs
+
+  nixops o c "modify" $ deploymentFiles cEnvironment cTarget cElements
+
+  nixops o c "deploy" [ "--max-concurrent-copy", "50", "-j", "4" ]
+  echo "Done."
+
+destroy :: Options -> NixopsConfig -> IO ()
+destroy o c@NixopsConfig{..} = do
+  printf ("Destroying cluster "%s%"\n") cName
+  nixops (o { oConfirm = True }) c "destroy" []
+  echo "Done."
+
+delete :: Options -> NixopsConfig -> IO ()
+delete o c@NixopsConfig{..} = do
+  printf ("Un-defining cluster "%s%"\n") cName
+  nixops (o { oConfirm = True }) c "delete" []
+  echo "Done."
+
+fromscratch :: Options -> NixopsConfig -> IO ()
+fromscratch o c = do
+  destroy o c
+  delete o c
+  create o c
+  deploy o c
+
+
+-- * Building
+--
+generateGenesis :: Options -> NixopsConfig -> IO ()
+generateGenesis o c = do
+  let cardanoSLSrcSpecFile = "cardano-sl-src.json"
+      cardanoSLDir         = "cardano-sl"
+  bs    <- BL.readFile cardanoSLSrcSpecFile
+  let NixGitSource{..} = flip fromMaybe (AE.decode bs) $
+                         errorT $ format ("Version spec for 'cardano-sl' doesn't parse as NixGitSource: "%s) (T.pack cardanoSLSrcSpecFile)
+  printf ("Generating genesis using cardano-sl commit "%s%"\n") $ fromCommit rev
+  exists <- testpath cardanoSLDir
+  unless exists $
+    cmd o "git" ["clone", fromURL cardanoSlURL, "cardano-sl"]
+  cd "cardano-sl"
+  cmd o "git" ["fetch"]
+  cmd o "git" ["reset", "--hard", fromCommit rev]
+  cd ".."
+  export "M" "14"
```

Comment:
The `--node-limit` is a dev-only option, to avoid the need patching `deployments/cardano-nodes-config.nix` -- it is not intended to be used for genesis.  The `deployments/cardano-nodes-config.nix` should be still authoritative for cluster configuration.

If we want to remove the `M=14` hard-code, we would need to come up with another solution -- for a problem that was different from the one the `--node-limit` is designed to solve.

And it's gonna be harder.